nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -234,52 +240,295 @@ impl Builtin for StringPrototypeIterator {
 }
 
 impl StringPrototype {
-    fn at(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
-    }
-
-    fn char_at(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
-    }
-
-    fn char_code_at(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
-    }
-
-    fn code_point_at(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
-    }
-
-    fn concat(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
-    }
-
-    fn ends_with(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
+    fn at(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+        // 2. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
+        // 3. Let len be the length of S.
+        let len = s.utf16_len(agent);
+        // 4. Let relativeIndex be ? ToIntegerOrInfinity(pos).
+        let relative_index = to_integer_or_infinity(agent, args.get(0))?.into_i64(agent);
```

Comment:
**thought:** I really dislike the `ToIntegerOrInfinity` AO, as it can still return a number that not a safe integer but also one that is probably far larger than `i64::MAX`.

I doubt rounding down to `i64::MAX` as opposed to some ridiculous unsafe f64 integer ever matters in the spec AOs, but it's bothered me a time or two: "Can I actually safely turn this value into an i64 or not?"

Reply:
I think for strings this is fine, because the maximum UTF-16 length of JS strings is `Number.MAX_SAFE_INTEGER`, which is less than `i64::MAX`, so anything that can't be represented in an `i64` is definitely past the end of the string.

That said, I don't think we're checking anywhere that strings can't go past that length, especially since we compute the UTF-16 length lazily. Maybe we should cap the UTF-8 length of a string to `3 * MAX_SAFE_INTEGER` and panic when computing the mapping if we're out of bounds of a valid JS string.

Reply:
Sounds like a good idea for the future at least.